commit a45581b9139c7a97a413f068aef06fe04bdccf6d
Author: Kody A Kantor <kody@kkantor.com>
Date:   2019-04-10T18:30:17+00:00 (6 months ago)
    
    OS-7709 metaslab load/unload counter kstats leak memory
    Reviewed by: Jerry Jelinek <jerry.jelinek@joyent.com>
    Reviewed by: Patrick Mooney <patrick.mooney@joyent.com>
    Approved by: Jerry Jelinek <jerry.jelinek@joyent.com>

diff --git a/usr/src/uts/common/fs/zfs/metaslab.c b/usr/src/uts/common/fs/zfs/metaslab.c
index bcc705ac9e..98e1c7b98f 100644
--- a/usr/src/uts/common/fs/zfs/metaslab.c
+++ b/usr/src/uts/common/fs/zfs/metaslab.c
@@ -23,6 +23,7 @@
  * Copyright (c) 2011, 2018 by Delphix. All rights reserved.
  * Copyright (c) 2013 by Saso Kiselkov. All rights reserved.
  * Copyright (c) 2014 Integros [integros.com]
+ * Copyright 2019 Joyent, Inc.
  */
 
 #include <sys/zfs_context.h>
@@ -850,7 +851,10 @@ metaslab_group_passivate(metaslab_group_t *mg)
 	mg->mg_next = NULL;
 
 	if (mg->mg_kstat != NULL) {
+		metaslab_group_kstat_t *data = mg->mg_kstat->ks_data;
+
 		kstat_delete(mg->mg_kstat);
+		kmem_free(data, sizeof (metaslab_group_kstat_t));
 	}
 	mutex_destroy(&mg->mg_kstat_lock);
 }
